starrSeq/alt.go

```diff
@@ -0,0 +1,121 @@
+package starrSeq
+
+import (
+	"fmt"
+	"github.com/vertgenlab/gonomics/bed"
+	"github.com/vertgenlab/gonomics/interval"
+	"github.com/vertgenlab/gonomics/sam"
+	"strings"
+)
+
+func Alt(s ScStarrSeqSettings) {
+	var tree = make([]interval.Interval, 0)
+	var selectTree map[string]*interval.IntervalNode
+	var overlap []interval.Interval
+	var starrSeqRead Read
+	var starrSeqReadSlice []Read
+	var allConstructs, allCellTypes, umiBxSlice []string
+	var cluster, cbcUpdate string
+	var found bool
+	var multipleGems bool = false
+	var gemNumber int = 1
+
+	cellTypeMap := make(map[string]string)
+
+	if s.ScAnalysis != "" {
+		ck := ReadClusterKey(s.ScAnalysis)
+		for _, i := range ck {
+			allCellTypes = append(allConstructs, i.Cluster)
+			cellTypeMap[i.Bx] = i.Cluster
+		}
+	}
+
+	bedEntries := bed.Read(s.AltMapping)
+	for _, i := range bedEntries {
+		tree = append(tree, i)
+		allConstructs = append(allConstructs, i.Name)
+	}
+	selectTree = interval.BuildTree(tree)
+
+	files := strings.Split(s.InFile, ",")
+	if len(files) > 1 {
+		multipleGems = true
+	}
+	for _, bam := range files {
+		inChan, _ := sam.GoReadToChan(bam)
+		for i := range inChan {
+			//filter sam file for some basic quality metrics
+			if !sam.FilterByQuality(i, 1) {
+				continue
+			}
+			if !sam.FilterByCigar(i, "starrSeqIntrons") {
+				continue
+			}
+			overlap = interval.Query(selectTree, i, "any")
+			if len(overlap) != 1 {
+				continue
+			}
+			cbc, cbcFound, _ := sam.QueryTag(i, "CB")
+			umi, umiFound, _ := sam.QueryTag(i, "UB")
+			if !umiFound {
+				continue
+			}
+			if multipleGems && cbcFound {
+				cbcUpdate = fmt.Sprintf("%s_%d", cbc.(string), gemNumber)
+				cluster, found = cellTypeMap[cbcUpdate]
+				umiBxSlice = append(umiBxSlice, fmt.Sprintf("%s_%s", umi.(string), cbcUpdate))
+			} else if cbcFound {
+				cbcUpdate = cbc.(string)
+				cluster, found = cellTypeMap[cbcUpdate]
+				umiBxSlice = append(umiBxSlice, fmt.Sprintf("%s_%s", umi.(string), cbcUpdate))
+			}
+			switch {
+			case !cbcFound:
+				starrSeqRead = Read{Bx: "NA", UMI: umi.(string), Cluster: "undefined", Construct: overlap[0].(bed.Bed).Name}
+			case cbcFound && found:
+				starrSeqRead = Read{Bx: cbcUpdate, UMI: umi.(string), Cluster: cluster, Construct: overlap[0].(bed.Bed).Name}
+			case cbcFound && !found:
+				starrSeqRead = Read{Bx: cbcUpdate, UMI: umi.(string), Cluster: "undefined", Construct: overlap[0].(bed.Bed).Name}
+			}
+			starrSeqReadSlice = append(starrSeqReadSlice, starrSeqRead)
+		}
+	}
+	collapsedUmiSlice := collapseUmi(starrSeqReadSlice)
```

Comment:
Alt does include collapseUmi, just with gonomics, not cellranger. Would it be helpful to users to change the cmd/starrSeqAnalysis.go usage to something like "collapseUmi by only removing identical cell barcode - UMI combinations through gonomics", rather than "by hand"?